<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>SingleFromCallableTest.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">RxJava_2_x$Unnamed.exec</a> &gt; <a href="index.source.html" class="el_package">io.reactivex.internal.operators.single</a> &gt; <span class="el_source">SingleFromCallableTest.java</span></div><h1>SingleFromCallableTest.java</h1><pre class="source lang-java linenums">/**
 * Copyright (c) 2016-present, RxJava Contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the &quot;License&quot;); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an &quot;AS IS&quot; BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
 * the License for the specific language governing permissions and limitations under the License.
 */

package io.reactivex.internal.operators.single;

import io.reactivex.Observer;
import io.reactivex.Single;
import io.reactivex.SingleObserver;
import io.reactivex.TestHelper;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.TestObserver;
import io.reactivex.plugins.RxJavaPlugins;
import io.reactivex.schedulers.Schedulers;
import org.junit.Test;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;

import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

<span class="fc" id="L39">public class SingleFromCallableTest {</span>

    @Test
    public void fromCallableValue() {
<span class="fc" id="L43">        Single.fromCallable(new Callable&lt;Integer&gt;() {</span>
            @Override public Integer call() throws Exception {
<span class="fc" id="L45">                return 5;</span>
            }
        })
<span class="fc" id="L48">            .test()</span>
<span class="fc" id="L49">            .assertResult(5);</span>
<span class="fc" id="L50">    }</span>

    @Test
    public void fromCallableError() {
<span class="fc" id="L54">        Single.fromCallable(new Callable&lt;Integer&gt;() {</span>
            @Override public Integer call() throws Exception {
<span class="fc" id="L56">                throw new UnsupportedOperationException();</span>
            }
        })
<span class="fc" id="L59">            .test()</span>
<span class="fc" id="L60">            .assertFailure(UnsupportedOperationException.class);</span>
<span class="fc" id="L61">    }</span>

    @Test
    public void fromCallableNull() {
<span class="fc" id="L65">        Single.fromCallable(new Callable&lt;Integer&gt;() {</span>
            @Override public Integer call() throws Exception {
<span class="fc" id="L67">                return null;</span>
            }
        })
<span class="fc" id="L70">            .test()</span>
<span class="fc" id="L71">            .assertFailureAndMessage(NullPointerException.class, &quot;The callable returned a null value&quot;);</span>
<span class="fc" id="L72">    }</span>

    @Test
    public void fromCallableTwice() {
<span class="fc" id="L76">        final AtomicInteger atomicInteger = new AtomicInteger();</span>

<span class="fc" id="L78">        Callable&lt;Integer&gt; callable = new Callable&lt;Integer&gt;() {</span>
            @Override
            public Integer call() throws Exception {
<span class="fc" id="L81">                return atomicInteger.incrementAndGet();</span>
            }
        };

<span class="fc" id="L85">        Single.fromCallable(callable)</span>
<span class="fc" id="L86">                .test()</span>
<span class="fc" id="L87">                .assertResult(1);</span>

<span class="fc" id="L89">        assertEquals(1, atomicInteger.get());</span>

<span class="fc" id="L91">        Single.fromCallable(callable)</span>
<span class="fc" id="L92">                .test()</span>
<span class="fc" id="L93">                .assertResult(2);</span>

<span class="fc" id="L95">        assertEquals(2, atomicInteger.get());</span>
<span class="fc" id="L96">    }</span>

    @SuppressWarnings(&quot;unchecked&quot;)
    @Test
    public void shouldNotInvokeFuncUntilSubscription() throws Exception {
<span class="fc" id="L101">        Callable&lt;Object&gt; func = mock(Callable.class);</span>

<span class="fc" id="L103">        when(func.call()).thenReturn(new Object());</span>

<span class="fc" id="L105">        Single&lt;Object&gt; fromCallableSingle = Single.fromCallable(func);</span>

<span class="fc" id="L107">        verifyZeroInteractions(func);</span>

<span class="fc" id="L109">        fromCallableSingle.subscribe();</span>

<span class="fc" id="L111">        verify(func).call();</span>
<span class="fc" id="L112">    }</span>

    @Test
    public void noErrorLoss() throws Exception {
<span class="fc" id="L116">        List&lt;Throwable&gt; errors = TestHelper.trackPluginErrors();</span>
        try {
<span class="fc" id="L118">            final CountDownLatch cdl1 = new CountDownLatch(1);</span>
<span class="fc" id="L119">            final CountDownLatch cdl2 = new CountDownLatch(1);</span>

<span class="fc" id="L121">            TestObserver&lt;Integer&gt; to = Single.fromCallable(new Callable&lt;Integer&gt;() {</span>
                @Override
                public Integer call() throws Exception {
<span class="fc" id="L124">                    cdl1.countDown();</span>
<span class="nc" id="L125">                    cdl2.await(5, TimeUnit.SECONDS);</span>
<span class="nc" id="L126">                    return 1;</span>
                }
<span class="fc" id="L128">            }).subscribeOn(Schedulers.single()).test();</span>

<span class="fc" id="L130">            assertTrue(cdl1.await(5, TimeUnit.SECONDS));</span>

<span class="fc" id="L132">            to.cancel();</span>

<span class="fc" id="L134">            int timeout = 10;</span>

<span class="pc bpc" id="L136" title="1 of 4 branches missed.">            while (timeout-- &gt; 0 &amp;&amp; errors.isEmpty()) {</span>
<span class="fc" id="L137">                Thread.sleep(100);</span>
            }

<span class="fc" id="L140">            TestHelper.assertUndeliverable(errors, 0, InterruptedException.class);</span>
        } finally {
<span class="fc" id="L142">            RxJavaPlugins.reset();</span>
        }
<span class="fc" id="L144">    }</span>

    @SuppressWarnings(&quot;unchecked&quot;)
    @Test
    public void shouldNotDeliverResultIfSubscriberUnsubscribedBeforeEmission() throws Exception {
<span class="fc" id="L149">        Callable&lt;String&gt; func = mock(Callable.class);</span>

<span class="fc" id="L151">        final CountDownLatch funcLatch = new CountDownLatch(1);</span>
<span class="fc" id="L152">        final CountDownLatch observerLatch = new CountDownLatch(1);</span>

<span class="fc" id="L154">        when(func.call()).thenAnswer(new Answer&lt;String&gt;() {</span>
            @Override
            public String answer(InvocationOnMock invocation) throws Throwable {
<span class="fc" id="L157">                observerLatch.countDown();</span>

                try {
<span class="nc" id="L160">                    funcLatch.await();</span>
<span class="fc" id="L161">                } catch (InterruptedException e) {</span>
                    // It's okay, unsubscription causes Thread interruption

                    // Restoring interruption status of the Thread
<span class="fc" id="L165">                    Thread.currentThread().interrupt();</span>
<span class="nc" id="L166">                }</span>

<span class="fc" id="L168">                return &quot;should_not_be_delivered&quot;;</span>
            }
        });

<span class="fc" id="L172">        Single&lt;String&gt; fromCallableObservable = Single.fromCallable(func);</span>

<span class="fc" id="L174">        Observer&lt;Object&gt; observer = TestHelper.mockObserver();</span>

<span class="fc" id="L176">        TestObserver&lt;String&gt; outer = new TestObserver&lt;String&gt;(observer);</span>

<span class="fc" id="L178">        fromCallableObservable</span>
<span class="fc" id="L179">                .subscribeOn(Schedulers.computation())</span>
<span class="fc" id="L180">                .subscribe(outer);</span>

        // Wait until func will be invoked
<span class="fc" id="L183">        observerLatch.await();</span>

        // Unsubscribing before emission
<span class="fc" id="L186">        outer.cancel();</span>

        // Emitting result
<span class="fc" id="L189">        funcLatch.countDown();</span>

        // func must be invoked
<span class="fc" id="L192">        verify(func).call();</span>

        // Observer must not be notified at all
<span class="fc" id="L195">        verify(observer).onSubscribe(any(Disposable.class));</span>
<span class="fc" id="L196">        verifyNoMoreInteractions(observer);</span>
<span class="fc" id="L197">    }</span>

    @Test
    public void shouldAllowToThrowCheckedException() {
<span class="fc" id="L201">        final Exception checkedException = new Exception(&quot;test exception&quot;);</span>

<span class="fc" id="L203">        Single&lt;Object&gt; fromCallableObservable = Single.fromCallable(new Callable&lt;Object&gt;() {</span>
            @Override
            public Object call() throws Exception {
<span class="fc" id="L206">                throw checkedException;</span>
            }
        });

<span class="fc" id="L210">        SingleObserver&lt;Object&gt; observer = TestHelper.mockSingleObserver();</span>

<span class="fc" id="L212">        fromCallableObservable.subscribe(observer);</span>

<span class="fc" id="L214">        verify(observer).onSubscribe(any(Disposable.class));</span>
<span class="fc" id="L215">        verify(observer).onError(checkedException);</span>
<span class="fc" id="L216">        verifyNoMoreInteractions(observer);</span>
<span class="fc" id="L217">    }</span>

    @Test
    public void disposedOnArrival() {
<span class="fc" id="L221">        final int[] count = { 0 };</span>
<span class="fc" id="L222">        Single.fromCallable(new Callable&lt;Object&gt;() {</span>
            @Override
            public Object call() throws Exception {
<span class="nc" id="L225">                count[0]++;</span>
<span class="nc" id="L226">                return 1;</span>
            }
        })
<span class="fc" id="L229">                .test(true)</span>
<span class="fc" id="L230">                .assertEmpty();</span>

<span class="fc" id="L232">        assertEquals(0, count[0]);</span>
<span class="fc" id="L233">    }</span>

    @Test
    public void disposedOnCall() {
<span class="fc" id="L237">        final TestObserver&lt;Integer&gt; to = new TestObserver&lt;Integer&gt;();</span>

<span class="fc" id="L239">        Single.fromCallable(new Callable&lt;Integer&gt;() {</span>
            @Override
            public Integer call() throws Exception {
<span class="fc" id="L242">                to.cancel();</span>
<span class="fc" id="L243">                return 1;</span>
            }
        })
<span class="fc" id="L246">                .subscribe(to);</span>

<span class="fc" id="L248">        to.assertEmpty();</span>
<span class="fc" id="L249">    }</span>

    @Test
    public void toObservableTake() {
<span class="fc" id="L253">        Single.fromCallable(new Callable&lt;Object&gt;() {</span>
            @Override
            public Object call() throws Exception {
<span class="fc" id="L256">                return 1;</span>
            }
        })
<span class="fc" id="L259">                .toObservable()</span>
<span class="fc" id="L260">                .take(1)</span>
<span class="fc" id="L261">                .test()</span>
<span class="fc" id="L262">                .assertResult(1);</span>
<span class="fc" id="L263">    }</span>

    @Test
    public void toObservableAndBack() {
<span class="fc" id="L267">        Single.fromCallable(new Callable&lt;Integer&gt;() {</span>
            @Override
            public Integer call() throws Exception {
<span class="fc" id="L270">                return 1;</span>
            }
        })
<span class="fc" id="L273">                .toObservable()</span>
<span class="fc" id="L274">                .singleOrError()</span>
<span class="fc" id="L275">                .test()</span>
<span class="fc" id="L276">                .assertResult(1);</span>
<span class="fc" id="L277">    }</span>
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.2.201808211720</span></div></body></html>